Improve Reference Column Iterable

mjendruk opened the pull request:
If a reference column references only a single chunk, the iterable will now resolve the type of the referenced column and will use the appropriate iterable to create its iterators.

Addresses #675

mrks commented on src/lib/storage/reference_column/reference_column_iterable.hpp:
```diff
@@ -17,74 +15,96 @@ class ReferenceColumnIterable : public ColumnIterable<ReferenceColumnIterable<T>
 
  private:
+  template <typename Functor>
+  void _referenced_with_iterators(const ChunkID chunk_id, const ChunkOffsetsList& chunk_offsets_list,
```
I don’t understand the name of this? The base is `with_iterators`, correct? What does “referenced” mean here?

mjendruk replied:
This function calls `with_iterators` of the referenced column.

mrks replied:
hmm, I don't like it, but I can't come up with anything better.

mjendruk replied:
`_with_iterators_of_referenced`?

mrks replied:
Yes :)

mrks commented on src/lib/storage/reference_column/reference_column_iterable.hpp:
```diff
@@ -17,74 +15,96 @@ class ReferenceColumnIterable : public ColumnIterable<ReferenceColumnIterable<T>
+        auto iterable = create_iterable_from_column<T>(referenced_column);
+        iterable.with_iterators(&chunk_offsets_list, [&](auto it, auto end) { functor(it, end); });
+      } else {
+        Fail("Reference column must not reference another reference column.");
```
Guard Clause?

mjendruk replied:
Doesn’t work with `if constexpr` I think.

mrzzzrm replied:
static_assert? :D

mjendruk replied:
Theoretically, this is never going to happen. I could replace the `else` block with something like `DebugAssert(is_reference_column, "Error message");`, but would this make it better?

mrks commented on src/lib/storage/reference_column/reference_column_iterable.hpp:
```diff
@@ -17,74 +15,96 @@ class ReferenceColumnIterable : public ColumnIterable<ReferenceColumnIterable<T>
-    const auto column_id = _column.referenced_column_id();
+    const auto pos_list = _column.pos_list();
+
+    const auto chunk_id_and_chunk_offsets_list = to_chunk_id_and_chunk_offsets_list(*pos_list);
```
Could you add a test for this new feature? Also, do the existing tests cover the case where the chunk id is not unique?

mjendruk replied:
You mean a test for this function? The tests should cover this case because it happens whenever a reference column created by a join operator is accessed.

mrks replied:
for everything. I wouldn't rely on unrelated tests to cover this. We now have a shortcut and relevant tests should make sure to cover both the fast and the slow path.

mrzzzrm commented on src/lib/storage/column_iterables/chunk_offset_mapping.cpp:
```diff
@@ -18,4 +21,34 @@ ChunkOffsetsByChunkID split_pos_list_by_chunk_id(const PosList& pos_list) {
   return chunk_offsets_by_chunk_id;
 }
 
+std::optional<std::pair<ChunkID, ChunkOffsetsList>> to_chunk_id_and_chunk_offsets_list(const PosList& pos_list) {
```
I love how the line comments in this functions increase it readability, even though they just rephrase the code below. Interesting...

mjendruk replied:
So is this good or bad? I’m not a fan of stating the obvious, but usually I’m asked to add more comments.

mrzzzrm replied:
Not sure tbh. I'd probably leave most of them out, but as I said they made things better for my first reading of this function. As part of this PR I'd just leave them in, we can discuss commenting strategies at another occasion.

mrzzzrm commented on src/lib/storage/reference_column/reference_column_iterable.hpp:
```diff
@@ -17,74 +15,96 @@ class ReferenceColumnIterable : public ColumnIterable<ReferenceColumnIterable<T>
+  template <typename Functor>
+  void _referenced_with_iterators(const ChunkID chunk_id, const ChunkOffsetsList& chunk_offsets_list,
+                                  const Functor& functor) const {
+    const auto column_id = _column.referenced_column_id();
```
when writing code in the context of ReferenceColumns, I'd prefer it if the prefixes "referenced" and "referencing"/"reference" would be used in variable names. But that's just opinion, feel free to ignore

mjendruk commented on src/lib/storage/column_iterables/chunk_offset_mapping.cpp:
```diff
@@ -18,4 +21,34 @@ ChunkOffsetsByChunkID split_pos_list_by_chunk_id(const PosList& pos_list) {
+    // Fail if PosList references more than one chunk
+    if (row_id.chunk_id != chunk_id) return std::nullopt;
+
+    chunk_offsets_list.push_back({chunk_offset, row_id.chunk_offset});
```
Maybe. Since the two values are integers, I guess the additional copies are removed by the compiler anyway. Unless we can verify the performance improvement, I wouldn’t change it.

mrzzzrm replied:
You're favouring last millennium's`push_back` over explicitly specifying you intend of constructing an object in place by using `emplace_back`? Wow, wouldn't have expected that.

mjendruk replied:
`emplace_back` has also disadvantages. It ignores constructors declared `explicit`. I think `emplace_back` is meant for those cases when the objects in the vector have complex (copy) constructors, so in order to avoid additional overhead, you can use `emplace_back`. For such primitive PODs as well as pointer types, I don’t think there is any benefit. But you can always prove me wrong!

mjendruk commented:
> @mrzzzrm I'm confused a bit. Didn't you intend to make the "references one Chunk" thing a flag of the PosList, set by the creating operator? The approach presented here needs to iterate the entire PosList before it can use the optimized access. That's twice the necessary accesses for a TableScan, isn't it?! Also, for multiple referenced Chunks, we still used operator[] AND type_cast and I thought you wanted to get rid of that as well?

Yes, but since we haven’t taken a decision how this flag will be stored, I added the optimisation without it and assumed the position list will reference only one chunk. Currently, whether we use the optimisation or not, the position list needs to be scanned and converted into one or several “chunk offsets lists“. This is how it is implemented in the table scan and this is how it was done even before I added the iterables. In this `ReferenceColumnIterable` implementation, if the position list references only one chunk, the position list is converted in the fastest way possible (I think) and then used to access the underlying column. In the other case, we still use the `operator[]` and `type_cast` yes. I had an idea how this could be done differently, then I had a conversation with @Bouncner and I decided that this is something that needs to be discussed in the group and needs to be evaluated properly.

mrzzzrm commented:
Okay, so this is just a temporary step then and ẃe'll stop using BaseColumn::operator[] after discussing things in person at some later point...

mrzzzrm commented:
Tests would be good, as mrks said.

adi64 commented:
I'd suggest to decide and merge #985 first. Then, I'd finish my "PosList with meta-information" PR and after that, we can take the fast path that is implemented here - without having the need to iterate over the PosList twice.

mrks commented:
Moved to Trello
